Modules/Core/Common/include/itkImageRange.h

```diff
@@ -0,0 +1,421 @@
+    // constructor that allow conversion from non-const to const iterator.
+    friend class QualifiedIterator<!VIsConst>;
+
+    // ImageRange is a friend, as it should be the only one that can
+    // directly use the private constructor of the iterator.
+    friend class ImageRange;
+
+    // Image type class that is either 'const' or non-const qualified, depending on QualifiedIterator and TImage.
+    using QualifiedImageType = typename std::conditional<VIsConst, const ImageType, ImageType>::type;
+
+    static constexpr bool IsImageTypeConst = std::is_const<QualifiedImageType>::value;
+
+    using QualifiedInternalPixelType = typename std::conditional<IsImageTypeConst, const InternalPixelType, InternalPixelType>::type;
+
+    // Pixel type class that is either 'const' or non-const qualified, depending on QualifiedImageType.
+    using QualifiedPixelType = typename std::conditional<IsImageTypeConst, const PixelType, PixelType>::type;
+
+    // The accessor of the image.
+    PixelAccessorType m_PixelAccessor;
+
+    QualifiedInternalPixelType* m_CurrentPixel = nullptr;
+
+    // Private constructor, used to create the begin and the end iterator of a range.
+    // Only used by its friend class ImageRange.
+    QualifiedIterator(
+      const PixelAccessorType& pixelAccessor,
+      QualifiedInternalPixelType* const currentPixel) ITK_NOEXCEPT
+      :
+    // Note: Use parentheses instead of curly braces to initialize data members,
+    // to avoid AppleClang 6.0.0.6000056 compilation error, "no viable conversion..."
+    m_PixelAccessor(pixelAccessor),
+    m_CurrentPixel{ currentPixel }
+    {
+    }
+
+  public:
+    // Types conforming the iterator requirements of the C++ standard library:
+    using difference_type = std::ptrdiff_t;
+    using value_type = PixelType;
+    using reference = QualifiedPixelType&;
+    using pointer = QualifiedPixelType*;
+    using iterator_category = std::random_access_iterator_tag;
+
+
+    /** Default-constructor, as required for any C++11 Forward Iterator. Offers
+     * the guarantee added to the C++14 Standard: "value-initialized iterators
+     * may be compared and shall compare equal to other value-initialized
+     * iterators of the same type."
+     */
+    QualifiedIterator() = default;
+
+    /** Constructor that allows implicit conversion from non-const to const
+     * iterator. Also serves as copy-constructor of a non-const iterator.  */
+    QualifiedIterator(const QualifiedIterator<false>& arg) ITK_NOEXCEPT
+      :
+      m_CurrentPixel{ arg.m_CurrentPixel },
+      // Note: Use parentheses instead of curly braces to initialize data members,
+      // to avoid AppleClang 6.0.0.6000056 compilation error, "no viable conversion..."
+      m_PixelAccessor(arg.m_PixelAccessor)
+    {
+    }
+
+
+    /**  Returns a reference to the current pixel. */
+    reference operator*() const ITK_NOEXCEPT
+    {
+      return m_PixelAccessor.Get(*m_CurrentPixel);
```

Comment:
@blowekamp Can you please check if I'm using the _pixel accessor_ correctly, here? The accessor is from image->GetPixelAccessor(), is that the right one?

Reply:
See also my Discourse post ["PixelAccessor or NeighborhoodAccessor?" ](https://discourse.itk.org/t/pixelaccessor-or-neighborhoodaccessor/1441 )